File: impilo/src/main/java/org/fhi360/plugins/impilo/domain/entities/Refill.java
package org.fhi360.plugins.impilo.domain.entities;

import com.blazebit.persistence.view.*;
import io.github.jbella.snl.core.api.domain.Organisation;
import io.github.jbella.snl.core.api.id.UUIDV7;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;
import org.fhi360.plugins.impilo.domain.providers.RefillBarcodeSubqueryProvider;

import java.time.LocalDate;
import java.util.UUID;

/**
 * The Refill class represents a refill of medication for a patient at a specific date, with various attributes and
 * associations.
 */
@Entity
@Getter
@Setter
@Table(name = "imp_refill")
public class Refill {
    @Id
    @UUIDV7
    private UUID id;

    private LocalDate date;

    private LocalDate dateNextRefill;

    private String regimen;

    private Integer qtyPrescribed;

    private Integer qtyDispensed;

    private Boolean missedDose;

    private Boolean adverseIssues;

    private Boolean synced = false;

    private UUID reference;

    private String batchIssuanceId;

    @ManyToOne
    @NotNull
    private Patient patient;

    @ManyToOne
    @NotNull
    private Organisation organisation;

    @EntityView(Refill.class)
    @CreatableEntityView
    public interface CreateView {
        @IdMapping
        UUID getId();

        @NotNull
        LocalDate getDate();

        void setDate(LocalDate date);

        LocalDate getDateNextRefill();

        void setDateNextRefill(LocalDate date);

        @NotNull
        String getRegimen();

        void setRegimen(String regimen);

        @NotNull
        @Min(0)
        Integer getQtyPrescribed();

        void setQtyPrescribed(Integer qty);

        @NotNull
        @Min(0)
        Integer getQtyDispensed();

        void setQtyDispensed(Integer qty);

        Boolean getMissedDose();

        void setMissedDose(Boolean missedDose);

        Boolean getAdverseIssues();

        void setAdverseIssues(Boolean adverseIssues);

        Boolean getSynced();

        void setSynced(Boolean synced);

        UUID getReference();

        void setReference(UUID reference);

        String getBatchIssuanceId();

        void setBatchIssuanceId(String batchIssuanceId);

        Patient.IdView getPatient();

        void setPatient(Patient.IdView patient);

        Organisation.IdView getOrganisation();

        void setOrganisation(Organisation.IdView organisation);

        @PostCreate
        default void postCreate() {
            setSynced(false);
        }
    }

    @EntityView(Refill.class)
    @UpdatableEntityView
    public interface UpdateView extends CreateView {
        void setId(UUID id);
    }

    @EntityView(Refill.class)
    public record ListView(LocalDate date, LocalDate dateNextRefill, String regimen, Integer qtyDispensed,
                           Integer qtyPrescribed) {

    }

    @EntityView(Refill.class)
    public record SyncView(LocalDate date, LocalDate dateNextRefill, String regimen, Integer qtyDispensed,
                           Integer qtyPrescribed, Patient.IdView patient, String batchIssuanceId,
                           Boolean missedDose, Boolean adverseIssues,
                           @MappingSubquery(RefillBarcodeSubqueryProvider.class)
                           String barcode) {

    }
}

File: impilo-mobile/lib/backend/floor/entities/refill.dart
import 'package:floor/floor.dart';
import 'package:impilo/flutter_flow/flutter_flow_util.dart';

@entity
class Refill {
  @PrimaryKey(autoGenerate: true)
  int? id;
  DateTime date;
  String regimen;
  String patientId;
  int quantityPrescribed;
  int quantityDispensed;
  DateTime dateNextRefill;
  bool synced;
  bool? missedDoses;
  bool? adverseIssues;
  String? barcode;
  String? batchIssuanceId;

  Refill(
      this.id,
      this.date,
      this.regimen,
      this.patientId,
      this.quantityPrescribed,
      this.quantityDispensed,
      this.dateNextRefill,
      this.missedDoses,
      this.adverseIssues,
      this.barcode,
      this.batchIssuanceId,
      this.synced);

  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String().substring(0, 10),
      'patient': {'id': patientId},
      'dateNextRefill': dateNextRefill.toIso8601String().substring(0, 10),
      'missedDoses': missedDoses,
      'adverseIssues': adverseIssues,
      'regimen': regimen,
      'batchIssuanceId': batchIssuanceId,
      'qtyPrescribed': quantityPrescribed,
      'qtyDispensed': quantityDispensed,
      'organisation': {'id': FFAppState().code}
    };
  }

  factory Refill.fromJson(Map<String, dynamic> row) => Refill(
      null,
      DateTime.parse(row['date']),
      row['regimen'],
      row['patient']['id'],
      row['qtyPrescribed'],
      row['qtyDispensed'],
      DateTime.parse(row['dateNextRefill']),
      row['missedDoses'],
      row['adverseIssues'],
      row['barcode'],
      row['batchIssuanceId'],
      true);
}
